### ISSUE TITLE
[Bug]: [Dynamic] Searched results is not flitered as specified expressions for dynamic field enabled collection 

### ISSUE THREAD

**Issue by binbinlv** [milvus-io/milvus#24514]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: master latest
- Deployment mode(standalone or cluster):both
- MQ type(rocksmq, pulsar or kafka):    all
- SDK version(e.g. pymilvus v2.0.0rc2): 2.4.0.dev48
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others:
```

### Current Behavior

 Searched results is not flitered as specified expressions for dynamic field enabled collection 

Here the expected result should be 997, while the actual result is 1000
```
res = collection.search(vectors[:nq], "float_vector", default_search_params, limit, "float not in [1, 2, 3]")
print(len(res[0].ids))
```

### Expected Behavior

Searched results is flitered as specified expressions for dynamic field enabled collection 

### Steps To Reproduce

```python
from pymilvus import CollectionSchema, FieldSchema
from pymilvus import Collection
from pymilvus import connections
from pymilvus import DataType
from pymilvus import Partition
import numpy as np
connections.connect()

int64_field = FieldSchema(name="int64", dtype=DataType.INT64, is_primary=True)
dim = 128
float_vector = FieldSchema(name="float_vector", dtype=DataType.FLOAT_VECTOR, dim=dim)
schema = CollectionSchema(fields=[int64_field, float_vector], enable_dynamic_field=True, auto_id=True)

collection = Collection("hello_milvus_binbin1", schema)
nb = 1000
rng = np.random.default_rng(seed=19530)
rows = [{"float": i*1.0, "varchar": str(i), "json_field": {"number": i, "string": str(i), "bool": bool(i), "list": [j for j in range(0, i)]}, "float_vector": rng.random((1, dim))[0]} for  i in range(0, nb)]
collection.insert(rows)
index_param = {"index_type": "IVF_FLAT", "metric_type": "L2", "params": {"nlist": 100}}
collection.create_index("float_vector", index_param)
collection.load()
default_nq = 10
import random
vectors = [[random.random() for _ in range(dim)] for _ in range(default_nq)]

default_search_params = {"metric_type": "L2", "params": {"nprobe": 10}}
limit = 1000
nq = 1
res = collection.search(vectors[:nq], "float_vector", default_search_params, limit, "float not in [1, 2, 3]")
print(len(res[0].ids))
```

**Comment by xiaofan-luan** [comment]
> ### Is there an existing issue for this?
> * [x]  I have searched the existing issues
> 
> ### Environment
> ```gfm
> - Milvus version: master latest
> - Deployment mode(standalone or cluster):both
> - MQ type(rocksmq, pulsar or kafka):    all
> - SDK version(e.g. pymilvus v2.0.0rc2): 2.4.0.dev48
> - OS(Ubuntu or CentOS): 
> - CPU/Memory: 
> - GPU: 
> - Others:
> ```
> 
> ### Current Behavior
> Searched results is not flitered as specified expressions for dynamic field enabled collection
> 
> Here the expected result should be 997, while the actual result is 1000
> 
> ```
> res = collection.search(vectors[:nq], "float_vector", default_search_params, limit, "float not in [1, 2, 3]")
> print(len(res[0].ids))
> ```
> 
> ### Expected Behavior
> Searched results is flitered as specified expressions for dynamic field enabled collection
> 
> ### Steps To Reproduce
> ```python
> from pymilvus import CollectionSchema, FieldSchema
> from pymilvus import Collection
> from pymilvus import connections
> from pymilvus import DataType
> from pymilvus import Partition
> import numpy as np
> connections.connect()
> 
> int64_field = FieldSchema(name="int64", dtype=DataType.INT64, is_primary=True)
> dim = 128
> float_vector = FieldSchema(name="float_vector", dtype=DataType.FLOAT_VECTOR, dim=dim)
> schema = CollectionSchema(fields=[int64_field, float_vector], enable_dynamic_field=True)
> 
> collection = Collection("hello_milvus_binbin1", schema)
> nb = 1000
> rng = np.random.default_rng(seed=19530)
> rows = [{"float": i*1.0, "varchar": str(i), "json_field": {"number": i, "string": str(i), "bool": bool(i), "list": [j for j in range(0, i)]}, "float_vector": rng.random((1, dim))[0]} for  i in range(0, nb)]
> collection.insert(rows)
> index_param = {"index_type": "IVF_FLAT", "metric_type": "L2", "params": {"nlist": 100}}
> collection.create_index("float_vector", index_param)
> collection.load()
> default_nq = 10
> import random
> vectors = [[random.random() for _ in range(dim)] for _ in range(default_nq)]
> 
> default_search_params = {"metric_type": "L2", "params": {"nprobe": 10}}
> limit = 1000
> nq = 1
> res = collection.search(vectors[:nq], "float_vector", default_search_params, limit, "float not in [1, 2, 3]")
> print(len(res[0].ids))
> ```
> 
> ### Milvus Log
> _No response_
> 
> ### Anything else?
> _No response_

 this is as expected?
Should it be float in [1.0, 2.0, 3.0]? 
I guess in didn't do implicit conversion

**Comment by binbinlv** [comment]
> float not in [1, 2, 3]

I think this is not as expected, If created float schema, then the result is right:
res = collection.search(vectors[:nq], "float_vector", default_search_params, limit, "float not in [1, 2, 3]")
997

So the behavior should keep the same?

```
from pymilvus import CollectionSchema, FieldSchema
from pymilvus import Collection
from pymilvus import connections
from pymilvus import DataType
from pymilvus import Partition
import numpy as np
connections.connect(host="10.102.9.26", port="19530")

int64_field = FieldSchema(name="int64", dtype=DataType.INT64, is_primary=True)
dim = 128
float_field = FieldSchema(name="float", dtype=DataType.FLOAT)
float_vector = FieldSchema(name="float_vector", dtype=DataType.FLOAT_VECTOR, dim=dim)
schema = CollectionSchema(fields=[int64_field, float_field, float_vector], enable_dynamic_field=True)

collection = Collection("hello_milvus_binbin1", schema)
nb = 1000
rng = np.random.default_rng(seed=19530)
rows = [{"int64": i, "float": i*1.0, "varchar": str(i), "json_field": {"number": i, "string": str(i), "bool": bool(i), "list": [j for j in range(0, i)]}, "float_vector": rng.random((1, dim))[0]} for  i in range(0, nb)]
collection.insert(rows)
index_param = {"index_type": "IVF_FLAT", "metric_type": "L2", "params": {"nlist": 100}}
collection.create_index("float_vector", index_param)
collection.load()
default_nq = 10
import random
vectors = [[random.random() for _ in range(dim)] for _ in range(default_nq)]

default_search_params = {"metric_type": "L2", "params": {"nprobe": 10}}
limit = 1000
nq = 1
res = collection.search(vectors[:nq], "float_vector", default_search_params, limit, "float not in [1, 2, 3]")
print(len(res[0].ids))
```

**Comment by yah01** [comment]
/assign @binbinlv 
Fixed with #24530 for master and #24539 for 2.2.0

**Comment by binbinlv** [comment]
Verified and fixed, so close this issue.

1. for 2.2.*
pymilvus: 2.2.9.dev30, milvus: 2.2.0-20230531-dbf07dfd
Results:
```
>>> res = collection.search(vectors[:nq], "float_vector", default_search_params, limit, "float not in [1, 2, 3]")
>>>
>>> print(len(res[0].ids))
997
>>> res = collection.search(vectors[:nq], "float_vector", default_search_params, limit, "float in [1, 2, 3] and float != 2")
>>> res[0].ids
[441846631035586972, 441846631035586974]
>>> print(len(res[0].ids))
2
>>> res = collection.search(vectors[:nq], "float_vector", default_search_params, limit, "float in [300/2, 900%40, -10*30+800, 2048/2%200, (100+200)*2]")
>>> print(len(res[0].ids))
5
```
2. for master:
pymilvus: 2.4.0.dev60, milvus: master-20230531-c7ec3b31
Results:
```
>>> res = collection.search(vectors[:nq], "float_vector", default_search_params, limit, "float not in [1, 2, 3]")
>>> print(len(res[0].ids))
997
>>> res = collection.search(vectors[:nq], "float_vector", default_search_params, limit, "float in [1, 2, 3] and float != 2")
>>> print(len(res[0].ids))
2
>>> res = collection.search(vectors[:nq], "float_vector", default_search_params, limit, "float in [300/2, 900%40, -10*30+800, 2048/2%200, (100+200)*2]")
>>> print(len(res[0].ids))
5
```
